1-Draft/RFCNNNN-Download-Cmdlet.md

```diff
@@ -0,0 +1,294 @@
+* Current Directory: `c:\images\`
+* Directory files: `gXnRf4J.jpg`
+* Current Time: 2018-03-31 08:11:29.632
+* File name: `c:\images\gXnRf4J.20180331081129632.jpg`
+
+Example 3:
+
+* Uri: `https://i.imgur.com/gXnRf4J.jpg`
+* Current Directory: `c:\images\`
+* Directory files: `gXnRf4J.jpg`, `gXnRf4J.20180331081129632.jpg`
+* Current Time: 2018-03-31 08:11:29.632
+* File name: `c:\images\gXnRf4J.20180331081129912.jpg`
+
+Example 4:
+
+* Uri: `https://i.imgur.com/`
+* Current Directory: `c:\images\`
+* Directory files: none
+* Current Time: 2018-03-31 08:11:29.632
+* File name: `c:\images\index.html`
+
+Example 5:
+
+* Uri: `https://i.imgur.com/`
+* Current Directory: `c:\images\`
+* Directory files: `index.html`
+* Current Time: 2018-03-31 08:11:29.632
+* File name: `c:\images\index.20180331081129632.html`
+
+Example 6:
+
+* Uri: `https://i.imgur.com/gXnRf4J.jpg?these=query&elements=will&be=ignored`
+* Current Directory: `c:\images\`
+* Directory files: none
+* Current Time: 2018-03-31 08:11:29.632
+* File name: `c:\images\gXnRf4J.jpg`
+
+Example 6:
+
+* Uri: `https://i.imgur.com/gXnRf4J>.jpg`
+* Current Directory: `c:\images\`
+* Directory files: none
+* Current Time: 2018-03-31 08:11:29.632
+* File name: `c:\images\gXnRf4J%3e.jpg`
+
+### PassThru Feature
+
+By default the cmdlet will return no output.
+The cmdlet is specialized for downloading.
+The cmdlet will perform no content encoding or decoding.
+The cmdlet only works with raw bytes.
+
+As the cmdlets primary function is to create a local file from a remote file,
+The only output it shall provide is the file it creates.
+
+By issuing the `-PassThru` switch,
+the cmdlet will return the equivalent of `Get-Item`.
+This will allow the cmdlet to pass the item to commands such as `Get-Content`
+where the user can specify what encoding features, if any, they require.
+
+This allows the cmdlet code to focus only on download features and optimizations
+without being coupled to the output ecosystem.
+For example, should the language adopt UTF-32 as the standard encoding,
+the cmdlet will not be coupled to that change.
+
+### AsJob Feature
+
+By default the cmdlet will run synchronously and block the current thread until completion.
+By supplying `-AsJob`, the user can run the cmdlet as a background job.
+As file downloads may take significant time,
+the ability to begin the download and perform other tasks should be native to the cmdlet.
+
+When supplied, the cmdlet will output a Job object.
+This feature will work in a manner similar to other cmdlets which implement it.
+
+### Multiple Protocol Support
+
+The cmdlet shall support multiple protocols.
+The cmdlet shall be extensible to support additional protocols.
+The protocol handlers shall be extensible to handle additional features.
+
+The extensibility is accomplished by interfaces.
+
+```csharp
+internal interface IInvokeDownloadProtocol
+{
+    Stream GetStream(PSCmdlet Cmdlet);
+    Uri GetUri();
+    void SetUri(Uri RequestUri);
```

Comment:
This interface might need an extra method `GetDynamicParameters(<some-parameter>)`. So if the protocol requires additional inputs from users, the implementation can extend the cmdlet syntax by defining dynamic parameters.